cd ~/Downloads/guileint-1.5.2/
diff -Naur /home/neil/Downloads/ /home/neil/Downloads/guileint-1.5.2/scheme.el
--- /home/neil/Downloads/scheme.el	Wed Aug 20 19:26:19 2003
+++ /home/neil/Downloads/guileint-1.5.2/scheme.el	Thu Oct  3 15:10:21 2002
@@ -157,6 +157,8 @@
   (setq parse-sexp-ignore-comments t)
   (make-local-variable 'lisp-indent-function)
   (set lisp-indent-function 'scheme-indent-function)
+  (make-local-variable 'scheme-mode-line-process) ;guileint added
+  (make-local-variable 'scheme-associated-process-buffer) ;guileint added
   (setq mode-line-process '("" scheme-mode-line-process))
   (set (make-local-variable 'imenu-case-fold-search) t)
   (setq imenu-generic-expression scheme-imenu-generic-expression)
@@ -170,6 +172,8 @@
           (font-lock-mark-block-function . mark-defun))))
 
 (defvar scheme-mode-line-process "")
+
+(defvar scheme-associated-process-buffer nil) ;guileint added
 
 (defvar scheme-mode-map nil
   "Keymap for Scheme mode.

Diff finished at Wed Aug 20 19:26:24
